packages/expo-sqlite/android/src/main/java/expo/modules/sqlite/SQLiteModule.kt

```diff
@@ -0,0 +1,170 @@
+        val bindArgs = convertParamsToStringArray(sqlQuery[1] as ArrayList<Any?>)
+        try {
+          if (isSelect(sql)) {
+            doSelectInBackgroundAndPossiblyThrow(sql, bindArgs, db)
+          } else { // update/insert/delete
+            if (readOnly) {
+              SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, ReadOnlyException())
+            } else {
+              doUpdateInBackgroundAndPossiblyThrow(sql, bindArgs, db)
+            }
+          }
+        } catch (e: Throwable) {
+          if (DEBUG_MODE) {
+            e.printStackTrace()
+          }
+          SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, 0, 0, e)
+        }
+      }
+      val data = pluginResultsToPrimitiveData(results)
+      promise.resolve(data)
+    } catch (e: Exception) {
+      promise.reject("SQLiteError", e)
+    }
+  }
+
+  @ExpoMethod
+  fun close(dbName: String, promise: Promise) {
+    DATABASES
+        .remove(dbName)
+        ?.close()
+    promise.resolve(null)
+  }
+
+  // do a update/delete/insert operation
+  private fun doUpdateInBackgroundAndPossiblyThrow(
+      sql: String,
+      bindArgs: Array<String?>?,
+      db: SQLiteDatabase
+  ): SQLitePluginResult {
+    return db.compileStatement(sql).use { statement ->
+      if (bindArgs != null) {
+        for (i in bindArgs.size downTo 1) {
+          if (bindArgs[i - 1] == null) {
+            statement.bindNull(i)
+          } else {
+            statement.bindString(i, bindArgs[i - 1])
+          }
+        }
+      }
+      if (isInsert(sql)) {
+        val insertId = statement.executeInsert()
+        val rowsAffected = if (insertId >= 0) 1 else 0
+        SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, rowsAffected, insertId, null)
+      } else if (isDelete(sql) || isUpdate(sql)) {
+        val rowsAffected = statement.executeUpdateDelete()
+        SQLitePluginResult(EMPTY_ROWS, EMPTY_COLUMNS, rowsAffected, 0, null)
+      } else {
+        // in this case, we don't need rowsAffected or insertId, so we can have a slight
+        // perf boost by just executing the query
+        statement.execute()
+        EMPTY_RESULT
+      }
+    }
+  }
+
+  // do a select operation
+  private fun doSelectInBackgroundAndPossiblyThrow(
+      sql: String,
+      bindArgs: Array<String?>,
+      db: SQLiteDatabase
+  ): SQLitePluginResult {
+    return db.rawQuery(sql, bindArgs).use { cursor ->
+      val numRows = cursor.count
+      if (numRows == 0) {
+        return EMPTY_RESULT
+      }
+      val numColumns = cursor.columnCount
+      val columnNames = cursor.columnNames
+      val rows: Array<Array<Any?>> = Array(numRows) { arrayOfNulls(numColumns) }
+      var i = 0
+      while (cursor.moveToNext()) {
+        val row = rows[i]
+        for (j in 0 until numColumns) {
+          row[j] = getValueFromCursor(cursor, j, cursor.getType(j))
+        }
+        rows[i] = row
+        i++
+      }
+      SQLitePluginResult(rows, columnNames, 0, 0, null)
+    }
```

Comment:
In contrary to the [`SQLiteStatemenet`](https://developer.android.com/reference/android/database/sqlite/SQLiteStatement) I believe [`SQLiteCursor`](https://developer.android.com/reference/android/database/sqlite/SQLiteCursor) should be manually closed? 🤔 
In JAVA we had:
```java
} finally {
  if (cursor != null) {
    cursor.close();
  }
```

Reply:
cursors also implement Closeable